File: csharp/patterns/src/PrototypePatterns/Structural/Composite/CompositePattern.cs
namespace PrototypePatterns.Structural.Composite;

public interface INode
{
    int Size();
}

public sealed class FileNode(int bytesSize) : INode
{
    public int Size() => bytesSize;
}

public sealed class FolderNode : INode
{
    private readonly List<INode> _children = [];

    public void Add(INode node) => _children.Add(node);

    public int Size() => _children.Sum(child => child.Size());
}

public static class CompositePattern
{
    public static int Demo()
    {
        var root = new FolderNode();
        root.Add(new FileNode(10));
        var nested = new FolderNode();
        nested.Add(new FileNode(5));
        nested.Add(new FileNode(15));
        root.Add(nested);
        return root.Size();
    }
}

File: java/patterns/src/main/java/org/prototypepatterns/structural/composite/CompositePattern.java
package org.prototypepatterns.structural.composite;

import java.util.ArrayList;
import java.util.List;

public final class CompositePattern {
    private CompositePattern() {
    }

    interface Node {
        int size();
    }

    static final class FileNode implements Node {
        private final int bytesSize;

        FileNode(int bytesSize) {
            this.bytesSize = bytesSize;
        }

        @Override
        public int size() {
            return bytesSize;
        }
    }

    static final class FolderNode implements Node {
        private final List<Node> children = new ArrayList<>();

        void add(Node node) {
            children.add(node);
        }

        @Override
        public int size() {
            return children.stream().mapToInt(Node::size).sum();
        }
    }

    public static int demo() {
        FolderNode root = new FolderNode();
        root.add(new FileNode(10));
        FolderNode nested = new FolderNode();
        nested.add(new FileNode(5));
        nested.add(new FileNode(15));
        root.add(nested);
        return root.size();
    }
}
